File: hotfix/src/main/java/com/xingshulin/singularity/core/Patch.java
package com.xingshulin.singularity.core;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Process;
import android.util.Log;
import com.xingshulin.singularity.utils.Configs;
import dalvik.system.DexClassLoader;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;

import static com.xingshulin.singularity.utils.ArrayUtils.concat;
import static com.xingshulin.singularity.utils.Configs.*;
import static com.xingshulin.singularity.utils.IOUtils.closeQuietly;
import static com.xingshulin.singularity.utils.IOUtils.save;
import static com.xingshulin.singularity.utils.ReflectionUtils.*;

public class Patch {
    public static final String TAG = "hotfix";
    static final String DOMAIN = "http://singularity.xingshulin.com";
    static final String KEY_URI = "uri";
    public static final String KEY_SHA = "sha1";

    public static void checkForUpdates(Context context) {
        if (isOnMainProcess(context)) {
            new DownloadThread(context).start();
        }
    }

    public static void configure(Context context, String token) {
        init(context, token);
        try {
            applyBasePatch(context);
            discoverAndApply(context);
        } catch (Exception e) {
            Log.w(TAG, "Configure failed, found error when applying patch: " + e.getMessage(), e);
        }
    }

    private static void applyBasePatch(Context context) {
        File apk = copyHelperApk(context, Configs.getDefaultPatchDir(context));
        loadPatch(apk.getAbsolutePath(), Configs.getDefaultPatchOptDir(context).getAbsolutePath());
    }

    private static void discoverAndApply(Context context) {
        File patchFile = getPatchFile(context);
        if (!patchFile.exists()) {
            return;
        }
        if (validatePatch(patchFile, getHotfixSha(context))) {
            apply(patchFile);
        } else {
            resetHotfixInfo(context);
            patchFile.delete();
        }
    }

    private static void apply(File patchFile) {
        loadPatch(patchFile.getAbsolutePath(), Configs.getPatchOptDir(patchFile).getAbsolutePath());
    }

    private static void loadPatch(String dexPath, String dexOpt) {
        Log.d(TAG, "loading patch file " + dexPath);
        DexClassLoader dexClassLoader = new DexClassLoader(dexPath, dexOpt, dexPath, getPathClassLoader());
        Object newDexElements = getDexElements(getPathList(dexClassLoader));
        Object baseDexElements = getDexElements(getPathList(getPathClassLoader()));
        Object concat = concat(newDexElements, baseDexElements);// Order is important
        Object pathList = getPathList(getPathClassLoader());
        setField(pathList, pathList.getClass(), "dexElements", concat);
    }

    private static ClassLoader getPathClassLoader() {
        return Patch.class.getClassLoader();
    }

    private static File copyHelperApk(Context context, File patchDir) {
        String apk = "patch_helper.apk";
        File copyTo = new File(patchDir, apk);
        if (copyTo.exists()) {
            return copyTo;
        }

        InputStream inputStream = null;
        try {
            inputStream = context.getAssets().open(apk);
            save(inputStream, copyTo);
            return copyTo;
        } catch (IOException e) {
            Log.w(TAG, e.getMessage(), e);
            return null;
        } finally {
            closeQuietly(inputStream);
        }
    }

    private static boolean isOnMainProcess(Context context) {
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningAppProcessInfo> infos = manager.getRunningAppProcesses();
        if (infos == null) {
            return false;
        }
        for (ActivityManager.RunningAppProcessInfo info : infos) {
            if (info.pid == Process.myPid()) {
                return context.getPackageName().equalsIgnoreCase(info.processName);
            }
        }
        return false;
    }

}

File: hotfix/src/main/kotlin/com/xingshulin/singularity/core/Patch.kt
package com.xingshulin.singularity.core

import android.content.Context
import android.util.Log
import com.xingshulin.singularity.utils.ArrayUtils.concat
import con.xingshulin.singularity.utils.*
import dalvik.system.DexClassLoader
import java.io.File

val TAG = "hotfix"
val KEY_URI = "uri"
val KEY_SHA = "sha1"
val DOMAIN = "http://singularity.xingshulin.com"

internal fun checkForUpdates(context: Context) {
    download(context)
}

fun configure(context: Context, token: String) {
    context.saveHotfixConfig("token", token)

    val apkDir = ensureSubDirExists(context.getHotfixRoot(), "default")
    val dexOptDir = ensureSubDirExists(apkDir, "dexOpt")
    val apk = copyHelperAPK(context, apkDir)

    loadPatch(apk, dexOptDir)
}

internal fun discoverAndApply(context: Context) {
    val patch = context.patchFile()
    if (!patch.exists()) return
    if (patch.isValidPatchFile(context)) {
        return apply(patch)
    } else {
        patch.delete()
    }
}

private fun apply(patch: File) {
    val dexOptDir = ensureSubDirExists(patch.parentFile, "dexOpt")
    loadPatch(patch, dexOptDir)
}

private fun loadPatch(dexFile: File, dexOptDir: File) {
    Log.d(TAG, "loading patch file ${dexFile.absolutePath}")
    val dexPath = dexFile.absolutePath
    val defaultDexOptPath = dexOptDir.absolutePath

    val dexClassLoader = DexClassLoader(dexPath, defaultDexOptPath, dexPath, getPathClassLoader())
    val baseDexElements = getDexElements(getPathList(getPathClassLoader()))
    val newDexElements = getDexElements(getPathList(dexClassLoader))

    val concat = concat(newDexElements, baseDexElements) //order is important!
    val pathList = getPathList(getPathClassLoader())
    setField(pathList, pathList.javaClass, "dexElements", concat)
}

private fun copyHelperAPK(context: Context, toDir: File): File {
    val apk = "patch_helper.apk"
    val input = context.assets.open(apk)
    val copyTo = File(toDir, apk)
    val out = copyTo.outputStream()
    input.copyTo(out)
    out.close()
    input.close()
    return copyTo
}

